src/MongoDB.Driver/ClientSessionHandle.cs

```diff
@@ -144,5 +244,162 @@ private TransactionOptions GetEffectiveTransactionOptions(TransactionOptions tra
 
             return new TransactionOptions(readConcern, readPreference, writeConcern);
         }
+
+        private WithTransactionStepResult<TResult> WithTransactionCallbackProcessing<TResult>(Func<IClientSessionHandle, TResult> callback, DateTime startTime, CancellationToken cancellationToken)
+        {
+            TResult result = default;
+            try
+            {
+                result = callback(this);
+                return new WithTransactionStepResult<TResult>(result, false);
+            }
+            catch (Exception ex)
+            {
+                if (IsInTransaction)
```

Comment:
Well, not 100%.
We have 4 possible transaction state on the DRIVER side: `Starting`, `InProgress`, `Committed`, `Aborted`.

The logic to calculate `IsInTransaction`:

    switch (_currentTransaction.State)
	{
		case CoreTransactionState.Aborted:
			return false;

		case CoreTransactionState.Committed:
			return _isCommitTransactionInProgress; // when retrying a commit we are temporarily "back in" the already committed transaction

		default:
			return true;
	}

So,  to have `IsInTransaction` with `true` value, a transaction state should have `Starting` or `InProgress` values which correspond for spec requirements.
The only one question is `_isCommitTransactionInProgress` value.
This variable always has `false` value except for the time when `CommitTransaction`(with retries on core side) is being processed. 
In this case this variable will have `true` value. I can be wrong, but as I see, from the SERVER point of view the transaction state will still be `InProcess` 
since transaction has been started to confirm once again. So, in some meaning this state is `InProgress` and not `Confirmed`.
We can discuss it too.

Reply:
I would rather see you match the spec than go into a subtle explanation of why IsInTransaction might happen to always result in the same truth value as what the spec says.

Reply:
Done